components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt

```diff
@@ -138,6 +143,34 @@ class BrowserIcons(
                 Logger.error("Could not install browser-icons extension", throwable)
             })
     }
+
+    /**
+     * Loads an icon asynchronously using [BrowserIcons] and then displays it in the [ImageView].
+     * If the view is detached from the window before loading is completed, then loading is cancelled.
+     */
+    fun loadIntoView(view: ImageView, request: IconRequest) = scope.async(Dispatchers.Main) {
```

Comment:
Since we are using the scope of `BrowserIcons` here, I wonder whether we could temporarily leak the view (and then its context/activity) by still having the coroutine around. We cancel when the View gets detached, but the cancellation itself may not happen immediately...? 🤔

Reply:
I'm not sure what the best way to proceed is here. 

After looking at the Kotlin docs, cancellation seems to be immediate as long as the function is in a suspended state. We could adjust `loadIcon` to check if the coroutine was cancelled.

We could store the view in a `WeakReference` and that might be enough to avoid memory leaks.  

I saw that Picasso implements LifecycleObserver, so you could optionally add it as an observer in the fragment and automatically cancel all jobs on destroy. Would that be a good pattern to follow?

Reply:
> After looking at the Kotlin docs, cancellation seems to be immediate as long as the function is in a suspended state. We could adjust loadIcon to check if the coroutine was cancelled.

The biggest issue here is probably the network request that may not really be cancelable as of now.

> We could store the view in a WeakReference and that might be enough to avoid memory leaks.

I think that would be a great idea.

> I saw that Picasso implements LifecycleObserver, so you could optionally add it as an observer in the fragment and automatically cancel all jobs on destroy. Would that be a good pattern to follow?

Oh, that's interesting.

We kind of should cover this with the attached state listener though; a destroy of the fragment should detach the views too.

And I assume we'd still run into the "cancel does not happen immediately" problem?

A `WeakReference` for the View sounds like the best option imo.